### PR TITLE
[libc++] Refactor __tuple_like and __pair_like

### PR DESCRIPTION
**Pull request by philnik777**
The exposition-only type trait `pair-like` includes `ranges::subrange`, but in every single case excludes `ranges::subrange` from the list. This patch introduces two new traits `__tuple_like_no_subrange` and `__pair_like_no_subrange`, which exclude `ranges::subrange` from the possible matches. `__pair_like` is no longer required, and thus removed. `__tuple_like` is implemented as `__tuple_like_no_subrange` or a `ranges::subrange` specialization.

### CODE REVIEW

**Review by ldionne — COMMENTED** [review]
TBH I am not certain about this refactor. Yes it makes sense from a certain perspective, but at the same time the Standard does define a `pair-like` concept (http://eel.is/c++draft/tuple.syn#concept:pair-like) so I find it better to define our implementation in terms of it.

**Comment by philnik777** [comment]
> TBH I am not certain about this refactor. Yes it makes sense from a certain perspective, but at the same time the Standard does define a `pair-like` concept (http://eel.is/c++draft/tuple.syn#concept:pair-like) so I find it better to define our implementation in terms of it.

I am aware of the exposition-only concept. My problem with implementing that 1:1 is that it creates a completely unnecessary dependency on `ranges::subrange`. That would be fine generally, but because it has concepts in the class signature even the forward declaration results in a quite significant dependency chain.

grafik

The highlighted part is `__fwd/subrange.h`. (FYI the box left of that is `__fwd/pair.h`)

**Comment by ldionne** [comment]
This is the first time I see this kind of output, and I think I am not able to interpret it properly. Where does that come from, is there any documentation or can you explain it to me? Sorry, but I kinda need that to properly understand what the benefit of this change is.

**Comment by philnik777** [comment]
> This is the first time I see this kind of output, and I think I am not able to interpret it properly. Where does that come from, is there any documentation or can you explain it to me?

You can get information about what the compiler is spending time on with `-ftime-trace`. You can get a visualization of that in chromium with `chrome://tracing/`. In this case it's just including `<__utility/pair.h>` (probably should have mentioned that) in C++23, which takes ~58ms, with 19ms spent on parsing `<__fwd/subrange.h>` and anything new it includes transitively.

> Sorry, but I kinda need that to properly understand (and quantify) what the benefit of this change is.

No problem.

File: libcxx/include/__tuple/tuple_like.h [hunk 1]
```diff
@@ -27,29 +24,15 @@ _LIBCPP_BEGIN_NAMESPACE_STD
...
 
 template <class _Tp>
-struct __tuple_like_impl<complex<_Tp>> : true_type {};
-
-#  endif
+concept __tuple_like = __tuple_like_no_subrange<_Tp> || __is_ranges_subrange_v<remove_cvref_t<_Tp>>;
 
-template <class _Tp>
-concept __tuple_like = __tuple_like_impl<remove_cvref_t<_Tp>>::value;
+// If the exposition-only type trait `pair-like` is required, you most likely want __pair_like_no_subrange.
```
Comment (mordante) [inline libcxx/include/__tuple/tuple_like.h:35]: Please explain why you most likely want this `__pair_like_no_subrange`. The commit message explains it, but in the future `__pair_like` might become useful, then this comment is misleading.

I also wonder whether it shouldn't be better to keep this in the `tuple_like` header. Putting it in a different header decreases its visibility.

**Review by mordante — COMMENTED** [review]
Thanks for working on compilation times!

Like @ldionne I also have some reservations about this change.

File: libcxx/include/__memory/uses_allocator_construction.h [hunk 2]
```diff
@@ -161,9 +157,7 @@ inline constexpr bool __convertible_to_const_pair_ref =
 #  if _LIBCPP_STD_VER >= 23
 template <class _Tp, class _Up>
 inline constexpr bool __uses_allocator_constraints =
-    __is_cv_std_pair<_Tp> &&
-    (__is_specialization_of_subrange<remove_cvref_t<_Up>>::value ||
-     (!__pair_like<_Up> && !__convertible_to_const_pair_ref<_Up>));
+    __is_cv_std_pair<_Tp> && !__pair_like_no_subrange<_Up> && !__convertible_to_const_pair_ref<_Up>;
```
Comment (ldionne) [inline libcxx/include/__memory/uses_allocator_construction.h:160]: After a lot of playing around with logical expressions, I can confirm before and after are equivalent.

File: libcxx/include/__ranges/subrange.h [hunk 3]
```diff
@@ -28,8 +28,9 @@
 #include <__ranges/enable_borrowed_range.h>
 #include <__ranges/size.h>
 #include <__ranges/view_interface.h>
-#include <__tuple/pair_like.h>
 #include <__tuple/tuple_element.h>
+#include <__tuple/tuple_like.h>
```
Comment (ldionne) [inline libcxx/include/__ranges/subrange.h:32]: Do you need `tuple_like` anywhere here?

**Review by ldionne — APPROVED** [review]
After going over this once more and understanding the tradeoffs better, I think this makes sense. At the end of the day the code is simpler too, it's just that the Standard is kinda weird for defining pair-like as tuple-like with size 2 when it really doesn't mean that (since that includes subrange, and we never want it).

**Pull request merged**